### PR TITLE
Change: Increase Internet Center sabotage duration of GLA Saboteur from 15000 to 60000 ms

### PR DESCRIPTION
**Pull request by xezon**
* Refers to #2097
* Related to #2355

This change increases the Internet Center sabotage duration of the GLA Saboteur from 15000 to 60000 ms.

This bumps the average gain factor on disabling 8 Hackers from 0.68 to 2.71. The gain factor can be lower or higher depending on the veterancy and amount of Hackers inside the Internet Center.

Saboteur build cost is 800.

### Math

Hackers give 5, 6, 8, 10 cash per tick, averaging at 7.25 cash per tick.

Internet Center tick is 1.6 seconds. 7.25 cash / 1.6 seconds = 4.53125 cash per second.

8 hackers * 4.53125 cash per second = 36.25 cash per second.

For fair sabotage value we use 2000, as much as original max cash gain.

2000 cash / 36.25 cash per second = 55 seconds.

We can round it up to 60 seconds.

## Original Internet Center sabotage

With a sabotage duration of 15 seconds.

| Hackers  | Sabotage cash value | Sabotage real value | Sabotage gain |
|----------|---------------------|---------------------|---------------|
| 8x Vet 0 | 375                 | -425                | 0.47 (-53%)   |
| 8x Vet 1 | 450                 | -350                | 0.56 (-44%)   |
| 8x Vet 2 | 600                 | -200                | 0.75 (-25%)   |
| 8x Vet 3 | 750                 | -50                 | 0.94 (-6%)    |

## Patched Internet Center sabotage

With a sabotage duration of 60 seconds.

| Hackers  | Sabotage cash value | Sabotage real value | Sabotage gain |
|----------|---------------------|---------------------|---------------|
| 8x Vet 0 | 1500                | +700                | 1.88 (+88%)   |
| 8x Vet 1 | 1800                | +1000               | 2.25 (+125%)  |
| 8x Vet 2 | 2400                | +1600               | 3.00 (+200%)  |
| 8x Vet 3 | 3000                | +2200               | 3.75 (+375%)  |

## Other things that can disable for reference

| Object                   | DisabledDuration |
|--------------------------|------------------|
| EMPPulseEffectSpheroid   |  30000           |
| LeafletContainer         |  20000           |
| EMPPatriotEffectSpheroid |  10000           |

## Patch China Economy buffs for reference

China has gotten a lot of buffs for its economy so far. Therefore this buff for GLA against China should be no major concern for China, but make the game content richer by giving the GLA Saboteur more valuable utility.

* #768
* #769
* #770
* #771
* #772
* #773
* #1533
* #1989
* #2027

# Rationale

In original game, it is not worthwhile to send the Saboteur into an Internet Center, because the maximum return for the sabotage duration of 15 seconds is lower than the actual cost of the Saboteur. Therefore the cash gain is negative. The only tangible advantage is the pausing of upgrade research and Satellite Hack spying for those 15 seconds.

This change increases the sabotage duration by +300% to achieve an average gain factor of 2.71 against a fully occupied Internet Center, which is similar to the max cash gain factor of 2.5 or 3.0 with the Supply Center (#2355).

The Internet Center can be easily defended from Saboteurs with the Mines upgrade.

### CODE REVIEW

**Comment by xezon** [comment]
After further investigation I have bumped the Internet Center sabotage duration from 15 to 60 seconds. This is more in line with the scale of #2355.

**Comment by MTKing4** [comment]
Can you evacuate hackers from Disabled Internet Centers? Because if you can this whole Advantage will be nullified.

**Comment by ExiLe-ZH** [comment]
> Can you evacuate hackers from Disabled Internet Centers? Because if you can this whole Advantage will be nullified.

well you said everything needs to be counterable, there you go ^^

On a real note though, first of all your hackers will be exposed when evacuating, secondly you will lose hacktime from various factors, like your reaction time, the walking and animations time etc. So is it even worth evacuating?

**Comment by MTKing4** [comment]
This is different, because the attacker is successful but is potentially not getting a reward or at least very little, but for example the cash steal already got their reward when they enter the building. You cannot return your $1000 after it was stolen

> On a real note though, first of all your hackers will be exposed when evacuating, secondly you will lose hacktime from various factors, like your reaction time, the walking and animations time etc. So is it even worth evacuating?

The question is, is it worth sabotaging? Because there's definitely a higher probability of near complete mitigation of the attack, making the enemy lose a small amount for relocating will not be worth the investment

That being said, it can be a nice combo tactic from the enemy, forcing hackers out with sabo then drop an anthrax bomb :D

Of course this all depends if my assumption about evacuating is true at all.

**Comment by xezon** [comment]
You can double check to verify, but I am pretty certain that the sabotaged structure will be disabled and it will not be possible to click buttons, just like you won't be able to build a tank on a sabotaged Arms Dealer.

The way to counter Saboteur is by upgrading Mines or placing appropriate units nearby.

shot_20230915_084252_2

**Comment by commy2** [comment]
That picture is from a Microwave tank, which behaves differently from e.g. EMP bomb.

Hackers outside the IC will also hack 16.7% slower, earning less money and experience.

What does Sabotage actually disable?
Does hacking an IC also reset / freeze / stop Sat Hack I and II?

**Comment by xezon** [comment]
Realistically no one will test this, so these questions cannot be answered at this time.

**Comment by ExiLe-ZH** [comment]
> The question is, is it worth sabotaging? Because there's definitely a higher probability of near complete mitigation of the attack, making the enemy lose a small amount for relocating will not be worth the investment

Exactly, it's high risk targetting the IC, because it's likely to be mined as it's a high value target. High risk should justify an high reward.

> That being said, it can be a nice combo tactic from the enemy, forcing hackers out with sabo then drop an anthrax bomb :D

Yep, or like an ambush whatever.
 
> Of course this all depends if my assumption about evacuating is true at all.

True, I got time tonight to test this.

**Comment by xezon** [comment]
Anything else to add?

**Comment by commy2** [comment]
I would like to know how this interacts with Sathack I & II.

**Comment by xezon** [comment]
Cash gain values corrected. Edit: Reverted, because it was correct before.

**Comment by xezon** [comment]
> I would like to know how this interacts with Sathack I & II.

Ok so here is what Saboteur does in Internet Center

# General, Hackers

* Disables upgrade buttons
* Pauses upgrade progress
* Pauses all hackers
* Does not disables evacuation

When evacuating Hacker(s), then Hacker(s) are disabled and cannot move. Hacker(s) can move after sabotage duration has ended.

https://github.com/TheSuperHackers/GeneralsGamePatch/assets/4720891/8da4e2a0-f0dc-404c-8824-9bf06ba31a86

# Satellite Hack I

The Satellite Hack I is currently bugged in Patch with #2416, but the Command Center reveal is disabled for the whole duration of the Internet Center sabotage.

https://github.com/TheSuperHackers/GeneralsGamePatch/assets/4720891/ddb90a8f-2c1b-4b71-8e08-5c8a34063b43

# Satellite Hack II (Patched Ability)

The Satellite Hack II map reveal is disabled during Internet Center sabotage - if it was activated prior the sabotage. During sabotage, Satellite Hack II cannot be activated. The Satellite Hack II cooldown timer will not be reset on sabotage.

https://github.com/TheSuperHackers/GeneralsGamePatch/assets/4720891/44d0c89c-ce36-406c-9b09-cfd86a0964f9

https://github.com/TheSuperHackers/GeneralsGamePatch/assets/4720891/ad82b908-861b-471a-b6b3-067921730052

**Comment by xezon** [comment]
Full information was appended to post above. Saboteur vs Internet Center is very effective. It does not just shutdown income generation, but also disables Satellite Hack I & II spying for the given duration. With the increased sabotage duration, this advantage is increased.

Are we good with this?

**Comment by xezon** [comment]
FiRe^ArM
> China's late game income is already bad. Doing it would worsen it even further.

**Comment by xezon** [comment]
> China's late game income is already bad. Doing it would worsen it even further.

China Economy and Internet Center buffs were added to opening post for reference.

**Comment by xezon** [comment]
Jundiyy
> 1 min does feel quite a long time, and since Hackers hack faster in the IC now it means a bigger loss than original (other than inf gen). Maybe 30 to 45 secs?

**Comment by xezon** [comment]
> 1 min does feel quite a long time, and since Hackers hack faster in the IC now it means a bigger loss than original (other than inf gen). Maybe 30 to 45 secs?

As outlined above, compared to the original Supply Center steal cash amount (1000 self reward + 1000 enemy loss = 2000), the fair Internet Center sabotage duration on average is 55 seconds. Therefore even 30 seconds are still not worth it. 45 seconds are not bad, but it is possible that players would still not risk the Saboteur for this time. The amount of success also depends on the amount and veterency of Hackers inside the Internet Center. If is filled with 8x Vet0 or 8x Vet1 Hackers, then it is still more valuable to sabotage the Supply Center for cash instead.

**Comment by xezon** [comment]
commy2
> I think the IC change is good.

**Comment by xezon** [comment]
Documentation added. Good.
